globus_sdk/transfer.py

```diff
@@ -1,9 +1,159 @@
+    chosen by other, future APIs.
+
+    A more complex and generic version of this generalization, as a
+    PaginatedResource class from which various API resources inherit is
+    possible, but it requires that individually defined API resources
+    "know" how to request the next page. We chose not to do this because of the
+    complexity and limited need for it at present.
+
+    When using this decorator, it is a good idea to name the arguments, even
+    though they are positionals, for greater clarity inline.
+    As in
+
+    >>> @paginated_resource(page_size=25, default_limit=25,
+    >>>                     max_total_results=250)
+    >>> def call_that_supports_paging(...):
+    >>>     ...
+
+    Expectations about Paginated Transfer API Resources:
+    - They support `limit` and `offset` query params, with `limit` being a
+      count of elements to return, and offset being an offset into the result
+      set (as opposed to a page number), 0-based
+    - They return a JSON result with `has_next_page` as a boolean key,
+      indicating whether or not there are more results available -- even if the
+      hard limit for the API forbids requesting these results
+    - Individual results are JSON objects inside of an array named `DATA` in
+      the returned JSON document
+    """
+
+    def wrapper(func):
+        """
+        Inner decorator returned by paginated_resource being called on its
+        arguments.
+        """
+        def _get_paginator_kwargs(**kwargs):
+            """
+            Small helper for tidiness. Pulls some values out of kwargs, as
+            desired.
+
+            TODO: Can functools make this easier for us?
+
+            We need to pull desired values out of kwargs so that we don't
+            disrupt the flow of original positional arguments to the wrapped
+            function, `func`
+            This is messiness because a function definition can't read as
+            >>> def wrapped_func(*args, a=1, **kwargs): ...
+
+            and if we do
+            >>> def wrapped_func(a=1, *args, **kwargs): ...
+
+            we'll get into trouble if we try to decorate a function with
+            positional arguments (because the first one will be assigned to
+            `a`, rather than whatever positional was used in the function
+            definition)
+            """
+            # pull out the paginate=True/False with a default of false
+            paginated = kwargs.get('paginate', False)
+
+            # limit for num results, with a default given by the decorator
+            # definition
+            limit = kwargs.get('limit', default_limit)
+
+            # offset should rarely be set, but we need to handle the case in
+            # which a caller wants to start their results at a given offset
+            # manually
+            offset = kwargs.get('offset', 0)
+
+            return paginated, limit, offset
+
+        def wrapped_func(*args, **kwargs):
```

Comment:
> You mean for function metadata preservation, right?

Yeah. It makes it less confusing if somebody is inspecting that method in a debugger or an interactive shell.